### PR TITLE
implement `TargetIdGenerator` in C++ for Firestore

### PR DESCRIPTION
**Pull request by zxu123**
### Discussion
part of the C++ port

### Testing
unit test ported from iOS test cases.

### API Changes
none

### CODE REVIEW

File: Firestore/core/src/firebase/firestore/core/target_id_generator.cc [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "Firestore/core/src/firebase/firestore/core/target_id_generator.h"
+
+#include <stdio.h>
+#include <atomic>
```
Comment (wilhuff) [inline Firestore/core/src/firebase/firestore/core/target_id_generator.cc]: As discussed in chat, the atomic changes in here are unnecessary. Target ID generators are never shared between threads.

File: Firestore/core/test/firebase/firestore/core/tmp/target_id_generator_test.cc [hunk 2]
```diff
@@ -0,0 +1,80 @@
+/*
```
Comment (wilhuff) [inline Firestore/core/test/firebase/firestore/core/tmp/target_id_generator_test.cc]: This file seems to have been added in error?

File: Firestore/core/src/firebase/firestore/core/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+cc_library(
+  firebase_firestore_core
+  SOURCES
+    target_id_generator.cc
```
Comment (wilhuff) [inline Firestore/core/src/firebase/firestore/core/CMakeLists.txt]: Please add headers too.

Comment (zxu123) [reply Firestore/core/src/firebase/firestore/core/CMakeLists.txt]: done

File: Firestore/core/src/firebase/firestore/core/target_id_generator.cc (hunk 1 above)
Comment (zxu123) [reply Firestore/core/src/firebase/firestore/core/target_id_generator.cc]: done

File: Firestore/core/test/firebase/firestore/core/tmp/target_id_generator_test.cc (hunk 2 above)
Comment (zxu123) [reply Firestore/core/test/firebase/firestore/core/tmp/target_id_generator_test.cc]: ack, reverted.

**Pull request merged**